core/tests/performance_framework/process_data.py

```diff
@@ -0,0 +1,128 @@
+use of the proxy, directly from the browser console.
+"""
+
+
+class PageSessionMetricsRetriever(object):
+    """Contains methods to process stats and provide performance metrics.
+
+    page_session_stats if a dictionary containing page load statistics or
+    HTTP Archive (HAR).
+    Refer: https://dvcs.w3.org/hg/webperf/raw-file/tip/specs/HAR/Overview.html
+
+    page_session_timings is a dictionary containing various timings associated
+    with a page load like request time, dom load time, etc.
+    Structure:
+        {
+          "webkitClearResourceTimings": {},
+          "memory": {...},
+          "timing": {
+            "domLoading": 1465470586738,
+            "fetchStart": 1465470583251,
+          }
+        }
+    """
+
+    def __init__(self, page_session_stats, page_session_timings):
+        self.page_session_stats = page_session_stats
+        self.page_session_timings = page_session_timings
+
+        self.page_load_timings = {}
+        if 'timing' not in self.page_session_timings:
+            raise ValueError('page_session_timings missing key: timing')
+
+        self.page_load_timings = page_session_timings['timing']
+
+    def get_request_count(self):
+        """Return no of requests made for the page load."""
+        request_count = 0
+
+        if 'log' in self.page_session_stats:
+            if 'entries' in self.page_session_stats['log']:
+                request_count = len(self.page_session_stats['log']['entries'])
+
+        if request_count == 0:
+            raise Exception('Total requests cannot be 0.')
+
+        return request_count
+
+    def get_total_page_size_bytes(self):
+        """Return total page size which includes the size of all resources
+        for a page.
+        """
+        total_size = 0
+
+        if 'log' in self.page_session_stats:
+            if 'entries' in self.page_session_stats['log']:
+                for entry in self.page_session_stats['log']['entries']:
+                    total_size += int(entry['response']['bodySize'])
+
+        if total_size <= 0:
+            raise Exception('Total page size should be positive.')
+
+        return total_size
+
+    def _get_duration_secs(self, event_end, event_initial):
+        # Check if timestamps are seconds or milliseconds.
+        # From: http://goo.gl/iHNYWx
+        if event_initial not in self.page_load_timings:
+            error_msg = 'page_load_timings missing key: %s' % event_initial
+            raise ValueError(error_msg)
+
+        initial_timestamp = self.page_load_timings[event_initial]
+
+        if event_end not in self.page_load_timings:
+            error_msg = 'page_load_timings missing key: %s' % event_end
+            raise ValueError(error_msg)
+
+        end_timestamp = self.page_load_timings[event_end]
+
+        # If milliseconds convert to seconds.
+        if len(str(initial_timestamp)) >= 13:
+            initial_timestamp /= 1000.0
+
+        if len(str(end_timestamp)) >= 13:
```

Comment:
It refers to the above comment thread that starts with "where the magic number 13 came from ?"